File: 0002-add-two-numbers/0002-add-two-numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        // approach: elementary addition
        
        int carry = 0;
        int top = 0;
        int bot = 0;
        
        ListNode dummy = new ListNode();
        ListNode curr = dummy;
        
        while (l1 != null || l2 != null) {
            
            if (l1 == null) {
                top = 0;
            } else {
                top = l1.val;
            }
            
            if (l2 == null) {
                bot = 0;
            } else {
                bot = l2.val;
            }
            
            int sum = carry + top + bot;
            carry = 0;
            
            if (sum > 9) {
                carry = sum / 10;
                sum = sum % 10;
            }
            
            curr.next = new ListNode(sum);
            curr = curr.next;

            if (l1 != null) {
                l1 = l1.next;
            }
            
            if (l2 != null) {
                l2 = l2.next;
            }
        }
        
        if (carry > 0) {
            curr.next = new ListNode(carry);
        }
        
        return dummy.next;
    }
}

File: 0002-add-two-numbers/0002-add-two-numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
    
        # approach: elementary addition, keep track of carry
    
        carry = 0
        head = ListNode()
        curr = head

        while l1 or l2:
            top = l1.val if l1 else 0
            bot = l2.val if l2 else 0

            result = carry + top + bot
            carry = 0

            if result > 9:
                result = result - 10
                carry = 1

            curr.next = ListNode(result)
            curr = curr.next
            
            if l1:
                l1 = l1.next
                
            if l2:
                l2 = l2.next

        if carry:
            curr.next = ListNode(carry)

        return head.next
        
        